// @nillion/secretvaults@0.1.0-rc.5/dist/lib.js
// src/builder.ts
import {
  InvocationBody,
  NilauthClient,
  NucTokenBuilder,
  PayerBuilder
} from "@nillion/nuc";

// src/common/blindfold.ts
import {
  allot,
  ClusterKey,
  encrypt,
  SecretKey,
  unify
} from "@nillion/blindfold";

// src/logger.ts
import pino from "pino";
var LOG_LEVEL_KEY = "NILLION_LOG_LEVEL";
var DEFAULT_LOG_LEVEL = "silent";
var VALID_LOG_LEVELS = [
  "fatal",
  "error",
  "warn",
  "info",
  "debug",
  "trace",
  "silent"
];
var safeStorage = {
  getItem: (key) => {
    try {
      return globalThis.localStorage?.getItem(key) ?? null;
    } catch {
      return null;
    }
  },
  setItem: (key, value) => {
    try {
      globalThis.localStorage?.setItem(key, value);
    } catch {
    }
  },
  removeItem: (key) => {
    try {
      globalThis.localStorage?.removeItem(key);
    } catch {
    }
  }
};
function isValidLogLevel(value) {
  return typeof value === "string" && VALID_LOG_LEVELS.includes(value);
}
function getInitialLogLevel() {
  let level = DEFAULT_LOG_LEVEL;
  const sources = [
    process.env?.[LOG_LEVEL_KEY],
    safeStorage.getItem(LOG_LEVEL_KEY),
    // @ts-ignore
    globalThis[LOG_LEVEL_KEY]
  ];
  for (const source of sources) {
    const levelAttempt = source?.toLowerCase();
    if (isValidLogLevel(levelAttempt)) {
      level = levelAttempt;
      break;
    }
  }
  console.log(`NILLION_LOG_LEVEL=${level}`);
  return level;
}
function getLogerForEnv() {
  const env = process.env.NODE_ENV ?? "production";
  if (env === "production") {
    return pino({
      level: getInitialLogLevel(),
      browser: {
        asObject: true
      },
      base: null
    });
  }
  return pino({
    level: getInitialLogLevel(),
    browser: {
      asObject: true
    },
    transport: {
      target: "pino-pretty",
      options: {
        colorize: true,
        levelFirst: true,
        translateTime: "SYS:h:MM:ss TT",
        ignore: "pid,hostname"
      }
    }
  });
}
var Log = getLogerForEnv();
function setLogLevel(level) {
  if (!isValidLogLevel(level)) {
    console.warn(`[Logger] Invalid log level: "${level}". Ignoring.`);
    return;
  }
  Log.level = level;
  safeStorage.setItem(LOG_LEVEL_KEY, level);
}
function getLogLevel() {
  return Log.level;
}
function clearStoredLogLevel() {
  safeStorage.removeItem(LOG_LEVEL_KEY);
}
if (typeof globalThis !== "undefined") {
  globalThis.__NILLION = {
    setLogLevel,
    getLogLevel,
    clearStoredLogLevel
  };
}

// src/common/blindfold.ts
async function toBlindfoldKey(options) {
  Log.debug(
    {
      hasExistingKey: "key" in options,
      operation: "operation" in options ? options.operation : "existing-key",
      clusterSize: options.clusterSize,
      useClusterKey: "useClusterKey" in options ? options.useClusterKey : false,
      hasSeed: "seed" in options && options.seed !== void 0
    },
    "Creating blindfold key"
  );
  if ("key" in options) {
    Log.debug({ keyType: options.key.constructor.name }, "Using existing key");
    return options.key;
  }
  const { operation, clusterSize } = options;
  const op = {
    [operation]: true
  };
  const threshold = "threshold" in options ? options.threshold : void 0;
  const cluster = { nodes: new Array(clusterSize).fill({}) };
  const useClusterKey = "useClusterKey" in options && options.useClusterKey;
  const useSeed = "seed" in options && options.seed !== void 0;
  const isClusterKey = useClusterKey || !useSeed && clusterSize > 1;
  const type = isClusterKey ? "ClusterKey" : "SecretKey";
  const key = isClusterKey ? await ClusterKey.generate(cluster, op, threshold) : await SecretKey.generate(
    cluster,
    op,
    threshold,
    "seed" in options ? options.seed : void 0
  );
  Log.debug(
    {
      key: type,
      operation,
      threshold,
      nodes: clusterSize
    },
    "Key generated"
  );
  return key;
}
async function conceal(key, data) {
  Log.debug(
    {
      keyType: key.constructor.name,
      dataKeys: Object.keys(data)
    },
    "Starting data concealment"
  );
  const encryptDeep = async (value) => {
    if (typeof value !== "object" || value === null) {
      return value;
    }
    if (Array.isArray(value)) {
      return Promise.all(value.map((e) => encryptDeep(e)));
    }
    const encryptedObj = {};
    for (const [oKey, oVal] of Object.entries(value)) {
      if (typeof oVal === "object" && oVal !== null) {
        if ("%allot" in oVal) {
          const plaintext = oVal["%allot"];
          encryptedObj[oKey] = {
            "%allot": await encrypt(key, plaintext)
          };
        } else {
          encryptedObj[oKey] = await encryptDeep(oVal);
        }
      } else {
        encryptedObj[oKey] = oVal;
      }
    }
    return encryptedObj;
  };
  const encryptedData = await encryptDeep(data);
  const shares = allot(encryptedData);
  Log.debug(
    { type: key.constructor.name, shares: shares.length },
    "Data concealed"
  );
  return shares;
}
async function reveal(key, shares) {
  const unified = await unify(key, shares);
  Log.debug(
    {
      type: key.constructor.name,
      keys: Object.keys(unified)
    },
    "Revealed data"
  );
  return unified;
}

// src/common/cluster.ts
async function executeOnCluster(nodes, operation) {
  Log.debug({ nodes: nodes.length }, "Executing cluster operation");
  const promises = nodes.map(async (client, index) => {
    const node = client.id.toString();
    Log.debug({ node, index }, "Starting node operation");
    try {
      const result = await operation(client, index);
      Log.debug({ node, index }, "Node operation completed");
      return [node, result];
    } catch (error) {
      Log.error({ node, index, error }, "Node operation failed");
      throw error;
    }
  });
  const results = await Promise.all(promises);
  Log.debug({ nodes: results.length }, "Cluster operation completed");
  return Object.fromEntries(results);
}
async function prepareConcealedRequest(options) {
  const { key, clients, body } = options;
  Log.debug(
    {
      key: key.constructor.name,
      nodes: clients.length,
      documents: body.data.length
    },
    "Preparing concealed data"
  );
  const concealedDocs = await Promise.all(
    body.data.map((d) => conceal(key, d))
  );
  if (concealedDocs.at(0)?.length !== clients.length) {
    Log.error(
      {
        shares: concealedDocs.at(0)?.length,
        nodes: clients.length
      },
      "Concealed shares count mismatch"
    );
    throw new Error("Concealed shares count mismatch.", { cause: options });
  }
  const sharesByNode = clients.map(
    (_, i) => concealedDocs.map((shares) => shares[i])
  );
  const pairs = clients.map((client, index) => {
    const payload = { ...body, data: sharesByNode[index] };
    return [client.id.toString(), payload];
  });
  Log.debug("Concealed data prepared");
  return Object.fromEntries(pairs);
}
function preparePlaintextRequest(options) {
  const { clients, body } = options;
  Log.debug({ nodes: clients.length }, "Preparing plaintext request");
  const pairs = clients.map(
    (c) => [c.id.toString(), { ...body }]
  );
  return Object.fromEntries(pairs);
}
function processPlaintextResponse(results, strategy = "first") {
  const values = Object.values(results);
  Log.debug(
    { nodes: values.length, strategy },
    "Processing plaintext response"
  );
  let index = 0;
  if (strategy === "random") {
    index = Math.floor(Math.random() * values.length);
  }
  const selected = values.at(index);
  if (selected === void 0) {
    Log.error({ resultsCount: values.length }, "No response to select");
    throw new Error("Failed to select a canonical response.", {
      cause: results
    });
  }
  Log.debug({ selectedIndex: index }, "Response selected");
  return selected;
}
async function processConcealedListResponse(options) {
  const { key, resultsByNode } = options;
  Log.debug(
    {
      key: key.constructor.name,
      nodes: Object.keys(resultsByNode).length
    },
    "Processing concealed list response"
  );
  const allShares = Object.values(resultsByNode).flatMap((r) => r.data);
  Log.debug({ totalShares: allShares.length }, "Flattened document shares");
  const groupedShares = allShares.reduce((acc, doc) => {
    const docId = doc._id;
    if (docId) {
      const group = acc.get(docId) ?? [];
      group.push(doc);
      acc.set(docId, group);
    }
    return acc;
  }, /* @__PURE__ */ new Map());
  Log.debug(
    { documentCount: groupedShares.size },
    "Grouped shares by document ID"
  );
  const revealPromises = Array.from(groupedShares.values()).map(
    (shares) => reveal(key, shares)
  );
  const revealed = await Promise.all(revealPromises);
  Log.debug(
    { revealedCount: revealed.length },
    "Documents revealed successfully"
  );
  return revealed;
}
async function processConcealedObjectResponse(options) {
  const { key, resultsByNode } = options;
  Log.debug(
    {
      key: key.constructor.name,
      nodes: Object.keys(resultsByNode).length
    },
    "Processing concealed object response"
  );
  const shares = Object.values(resultsByNode).map((response) => response.data);
  Log.debug({ shareCount: shares.length }, "Collected object shares");
  const revealed = await reveal(key, shares);
  Log.debug("Object revealed successfully");
  return revealed;
}

// src/base.ts
var SecretVaultBaseClient = class {
  _options;
  constructor(options) {
    this._options = options;
  }
  get id() {
    return this.did.toString();
  }
  /**
   * The DID of the keypair associated with this client.
   */
  get did() {
    return this._options.keypair.toDid();
  }
  /**
   * The array of underlying node clients for the cluster.
   */
  get nodes() {
    return this._options.clients;
  }
  /**
   * The keypair used by this client for signing.
   */
  get keypair() {
    return this._options.keypair;
  }
  /**
   * Retrieves information about each node in the cluster.
   */
  async readClusterInfo() {
    const result = await executeOnCluster(this.nodes, (c) => c.aboutNode());
    Log.info({ nodes: Object.keys(result).length }, "Cluster info retrieved");
    return result;
  }
};

// src/common/time.ts
function intoSecondsFromNow(seconds) {
  return Math.floor((Date.now() + seconds * 1e3) / 1e3);
}

// src/common/nuc-cmd.ts
import { Command } from "@nillion/nuc";
function createNucNamespace(prefix) {
  const base = prefix.split("/");
  return {
    root: new Command([...base]),
    create: new Command([...base, "create"]),
    read: new Command([...base, "read"]),
    update: new Command([...base, "update"]),
    delete: new Command([...base, "delete"]),
    execute: new Command([...base, "execute"])
  };
}
var NucCmd = {
  nil: {
    db: {
      root: new Command(["nil", "db"]),
      system: createNucNamespace("nil/db/system"),
      builders: createNucNamespace("nil/db/builders"),
      data: createNucNamespace("nil/db/data"),
      collections: createNucNamespace("nil/db/collections"),
      queries: createNucNamespace("nil/db/queries"),
      users: createNucNamespace("nil/db/users")
    }
  }
};

// src/common/types.ts
import { z } from "zod/v4";
var Uuid = z.uuid().brand();
var Did = z.string().regex(/^did:nil:([a-zA-Z0-9]{66})$/).brand();

// src/nildb/builder-client.ts
import { z as z10 } from "zod/v4";

// src/common/paths.ts
var NilDbEndpoint = {
  v1: {
    builders: {
      register: "/v1/builders/register",
      me: "/v1/builders/me"
    },
    data: {
      root: "/v1/data",
      find: "/v1/data/find",
      update: "/v1/data/update",
      delete: "/v1/data/delete",
      flushById: "/v1/data/:id/flush",
      tailById: "/v1/data/:id/tail",
      createOwned: "/v1/data/owned",
      createStandard: "/v1/data/standard"
    },
    queries: {
      root: "/v1/queries",
      byId: "/v1/queries/:id",
      run: "/v1/queries/run",
      runById: "/v1/queries/run/:id"
    },
    collections: {
      root: "/v1/collections",
      byId: "/v1/collections/:id",
      indexesById: "/v1/collections/:id/indexes",
      indexesByNameById: "/v1/collections/:id/indexes/:name"
    },
    system: {
      about: "/about",
      health: "/health",
      metrics: "/metrics",
      openApiJson: "/openapi.json",
      maintenanceStart: "/v1/system/maintenance/start",
      maintenanceStop: "/v1/system/maintenance/stop",
      logLevel: "/v1/system/log-level"
    },
    users: {
      me: "/v1/users/me",
      data: {
        root: "/v1/users/data",
        byId: "/v1/users/data/:collection/:document",
        aclById: "/v1/users/data/:collection/:document/acl",
        acl: {
          grant: "/v1/users/data/acl/grant",
          revoke: "/v1/users/data/acl/revoke"
        }
      }
    }
  }
};

// src/dto/builders.dto.ts
import { z as z3 } from "zod/v4";

// src/dto/common.ts
import { z as z2 } from "zod/v4";
var Name = z2.string().min(1).max(255);
var ApiSuccessResponse = (Schema) => z2.object({
  data: Schema
});
var ApiErrorResponse = z2.object({
  ts: z2.iso.datetime(),
  errors: z2.string().array()
});
var ByIdRequestParams = z2.object({
  id: z2.uuid()
});
var Acl = z2.object({
  grantee: Did,
  read: z2.boolean(),
  write: z2.boolean(),
  execute: z2.boolean()
});

// src/dto/builders.dto.ts
var RegisterBuilderRequest = z3.object({
  did: Did,
  name: Name
});
var RegisterBuilderResponse = z3.string();
var BuilderProfileDto = z3.object({
  _id: Did,
  _created: z3.iso.datetime(),
  _updated: z3.iso.datetime(),
  name: z3.string(),
  collections: z3.array(z3.uuid()),
  queries: z3.array(z3.uuid())
});
var ReadBuilderProfileResponse = ApiSuccessResponse(BuilderProfileDto);
var DeleteBuilderResponse = z3.string();
var UpdateBuilderProfileRequest = z3.object({
  name: Name
});
var UpdateBuilderProfileResponse = z3.string();

// src/dto/collections.dto.ts
import { z as z4 } from "zod/v4";
var CollectionDocumentDto = z4.object({
  owner: Did,
  type: z4.enum(["standard", "owned"]),
  name: z4.string(),
  schema: z4.record(z4.string(), z4.unknown())
});
var CollectionDocuments = z4.array(CollectionDocumentDto);
var ListCollectionsResponse = ApiSuccessResponse(CollectionDocuments);
var CreateCollectionIndexRequest = z4.object({
  collection: z4.uuid(),
  name: z4.string().min(4),
  keys: z4.array(
    z4.record(z4.string(), z4.union([z4.literal(1), z4.literal(-1)])).refine(
      (obj) => Object.keys(obj).length === 1,
      "Each object must have exactly one key: [{ _id: 1 }, { foo: -1 }]"
    )
  ),
  unique: z4.boolean(),
  ttl: z4.number().optional()
});
var CreateCollectionIndexResponse = z4.string();
var DropCollectionIndexParams = z4.object({
  id: z4.uuid(),
  name: z4.string().min(4).max(50)
});
var DropCollectionIndexResponse = z4.string();
var CreateCollectionRequest = z4.object({
  _id: z4.uuid(),
  type: z4.union([z4.literal("standard"), z4.literal("owned")]),
  name: z4.string().min(1),
  schema: z4.record(z4.string(), z4.unknown())
});
var CreateCollectionResponse = z4.string();
var DeleteCollectionRequestParams = z4.object({
  id: z4.uuid()
});
var DeleteCollectionResponse = z4.string();
var ReadCollectionMetadataRequestParams = z4.object({
  id: z4.uuid()
});
var CollectionIndexDto = z4.object({
  v: z4.number(),
  key: z4.record(z4.string(), z4.union([z4.string(), z4.number()])),
  name: z4.string(),
  unique: z4.boolean()
});
var CollectionMetadataDto = z4.object({
  _id: z4.uuid(),
  count: z4.number(),
  size: z4.number(),
  first_write: z4.iso.datetime(),
  last_write: z4.iso.datetime(),
  indexes: z4.array(CollectionIndexDto)
});
var ReadCollectionMetadataResponse = ApiSuccessResponse(
  CollectionMetadataDto
);

// src/dto/data.dto.ts
import { z as z6 } from "zod/v4";

// src/dto/users.dto.ts
import { z as z5 } from "zod/v4";
var AclDto = z5.object({
  grantee: Did,
  read: z5.boolean(),
  write: z5.boolean(),
  execute: z5.boolean()
});
var UserProfileData = z5.object({
  _id: Did,
  _created: z5.iso.datetime(),
  _updated: z5.iso.datetime(),
  logs: z5.array(
    z5.looseObject({
      op: z5.string(),
      collection: z5.uuid(),
      // present when op is "auth"
      acl: AclDto.optional()
    })
  ),
  data: z5.array(
    z5.object({
      collection: z5.uuid(),
      id: z5.uuid()
    })
  )
});
var ReadUserProfileResponse = ApiSuccessResponse(UserProfileData);
var ReadDataRequestParams = z5.object({
  collection: z5.uuid(),
  document: z5.uuid()
});
var OwnedDataDto = z5.looseObject({
  _id: z5.uuid(),
  _created: z5.iso.datetime(),
  _updated: z5.iso.datetime(),
  _owner: Did,
  _acl: z5.array(AclDto)
});
var ReadDataResponse = ApiSuccessResponse(OwnedDataDto);
var DataDocumentReference = z5.object({
  builder: Did,
  collection: z5.uuid(),
  document: z5.uuid()
});
var ListDataReferencesResponse = ApiSuccessResponse(
  z5.array(DataDocumentReference)
);
var ReadDataAclRequestParams = z5.object({
  collection: z5.uuid(),
  document: z5.uuid()
});
var ReadDataAccessResponse = ApiSuccessResponse(z5.array(AclDto));
var GrantAccessToDataRequest = z5.object({
  collection: z5.uuid(),
  document: z5.uuid(),
  acl: AclDto
});
var GrantAccessToDataResponse = z5.string();
var RevokeAccessToDataRequest = z5.object({
  grantee: Did,
  collection: z5.uuid(),
  document: z5.uuid()
});
var RevokeAccessToDataResponse = z5.string();
var DeleteDocumentRequestParams = z5.object({
  collection: z5.uuid(),
  document: z5.uuid()
});
var DeleteDocumentResponse = z5.string();
var UpdateUserDataRequest = z5.object({
  document: z5.uuid(),
  collection: z5.uuid(),
  update: z5.record(z5.string(), z5.unknown())
});

// src/dto/data.dto.ts
var CreateOwnedDataRequest = z6.object({
  owner: Did,
  collection: z6.uuid(),
  data: z6.array(z6.record(z6.string(), z6.unknown())).min(1),
  acl: AclDto
});
var CreateStandardDataRequest = z6.object({
  collection: z6.uuid(),
  data: z6.array(z6.record(z6.string(), z6.unknown())).min(1)
});
var CreateDataResponse = ApiSuccessResponse(
  z6.object({
    created: z6.array(z6.uuid()),
    errors: z6.array(
      z6.object({
        error: z6.string(),
        document: z6.unknown()
      })
    )
  })
);
var UpdateDataRequest = z6.object({
  collection: z6.uuid(),
  filter: z6.record(z6.string(), z6.unknown()),
  update: z6.record(z6.string(), z6.unknown())
});
var UpdateDataResponse = ApiSuccessResponse(
  z6.object({
    acknowledged: z6.boolean(),
    matched: z6.number().int().min(0),
    modified: z6.number().int().min(0),
    upserted: z6.number().int().min(0),
    upserted_id: z6.string().nullable()
  })
);
var FindDataRequest = z6.object({
  collection: z6.uuid(),
  filter: z6.record(z6.string(), z6.unknown())
});
var FindDataResponse = ApiSuccessResponse(
  z6.array(z6.record(z6.string(), z6.unknown()))
);
var DeleteDataRequest = z6.object({
  collection: z6.uuid(),
  filter: z6.record(z6.string(), z6.unknown()).refine((obj) => Object.keys(obj).length > 0, "Filter cannot be empty")
});
var DeleteDataResponse = ApiSuccessResponse(
  z6.object({
    acknowledged: z6.boolean(),
    deletedCount: z6.number().int().min(0)
  })
);
var FlushDataRequest = z6.object({
  collection: z6.uuid()
});
var FlushDataResponse = z6.string();
var DropDataResponse = ApiSuccessResponse(
  z6.object({
    acknowledged: z6.boolean(),
    deletedCount: z6.number().int().min(0)
  })
);
var DataSchemaByIdRequestParams = z6.object({
  id: z6.uuid()
});
var TailDataRequestParams = z6.object({
  id: z6.uuid()
});
var TailDataRequestQuery = z6.object({
  limit: z6.coerce.number().max(1e3).optional().default(25)
});
var TailDataResponse = ApiSuccessResponse(
  z6.array(z6.record(z6.string(), z6.unknown()))
);

// src/dto/queries.dto.ts
import { z as z7 } from "zod/v4";
var PATH_EXPRESSION = /^\$(\.[$a-zA-Z][a-zA-Z0-9-_]+(\[\d+])*)+$/;
var VariablePath = z7.string().transform((path) => PATH_EXPRESSION.exec(path)).refine((match) => match !== null, "invalid PATH").transform((match) => match[0]);
var QueryVariableValidator = z7.object({
  path: VariablePath,
  description: z7.string().optional()
});
var CreateQueryRequest = z7.object({
  _id: z7.uuid(),
  name: z7.string().min(1).max(100),
  collection: z7.uuid(),
  variables: z7.record(z7.string(), QueryVariableValidator),
  pipeline: z7.array(z7.record(z7.string(), z7.unknown()))
});
var CreateQueryResponse = z7.string();
var QueryDocumentResponse = z7.object({
  _id: z7.uuid(),
  _created: z7.iso.datetime(),
  _updated: z7.iso.datetime(),
  owner: Did,
  name: z7.string().min(1).max(100),
  collection: z7.uuid(),
  variables: z7.record(z7.string(), QueryVariableValidator),
  pipeline: z7.array(z7.record(z7.string(), z7.unknown()))
});
var ReadQueriesResponse = ApiSuccessResponse(
  z7.array(QueryDocumentResponse)
);
var ReadQueryResponse = ApiSuccessResponse(QueryDocumentResponse);
var DeleteQueryRequest = z7.object({
  id: z7.uuid()
});
var DeleteQueryResponse = z7.string();
var RunQueryRequest = z7.object({
  _id: z7.uuid(),
  variables: z7.record(z7.string(), z7.unknown())
});
var RunQueryResponse = ApiSuccessResponse(z7.uuid());
var RunQueryResultStatus = z7.enum([
  "pending",
  "running",
  "complete",
  "error"
]);
var ReadQueryRunByIdDto = z7.object({
  _id: z7.uuid(),
  query: z7.uuid(),
  status: RunQueryResultStatus,
  started: z7.iso.datetime().optional(),
  completed: z7.iso.datetime().optional(),
  result: z7.unknown().optional(),
  errors: z7.array(z7.string()).optional()
});
var ReadQueryRunByIdResponse = ApiSuccessResponse(ReadQueryRunByIdDto);

// src/nildb/base-client.ts
import { Did as NucDid } from "@nillion/nuc";
import { z as z9 } from "zod/v4";

// src/dto/system.dto.ts
import { z as z8 } from "zod/v4";
var ReadAboutNodeResponse = z8.object({
  started: z8.iso.datetime(),
  build: z8.object({
    time: z8.iso.datetime(),
    commit: z8.string(),
    version: z8.string()
  }),
  public_key: z8.string(),
  url: z8.string().url(),
  maintenance: z8.object({
    active: z8.boolean(),
    started_at: z8.iso.datetime()
  })
});
var NodeHealthCheckResponse = z8.literal("OK");

// src/nildb/base-client.ts
var NilDbBaseClientOptions = z9.object({
  about: ReadAboutNodeResponse,
  baseUrl: z9.string().min(15)
});
var NilDbBaseClient = class {
  #options;
  constructor(options) {
    this.#options = options;
  }
  get name() {
    return this.#options.about.public_key.slice(-4);
  }
  get id() {
    return NucDid.fromHex(this.#options.about.public_key);
  }
  /**
   * Handles error responses with consistent error information
   */
  handleErrorResponse(response, method, path, body) {
    throw new Error(`Request failed: ${method} ${path}`, {
      cause: {
        body,
        response,
        status: response.status,
        statusText: response.statusText
      }
    });
  }
  /**
   * Makes an authenticated request to the NilDb API
   */
  async request(options) {
    const { path, token, method = "GET", body, responseSchema } = options;
    const headers = {};
    if (token) {
      headers.Authorization = `Bearer ${token}`;
    }
    if (body) {
      headers["Content-Type"] = "application/json";
    }
    const endpoint = new URL(path, this.#options.baseUrl).toString();
    const response = await fetch(endpoint, {
      method,
      headers,
      ...body && { body: JSON.stringify(body) }
    });
    const contentType = response.headers.get("content-type") ?? "";
    const status = response.status;
    if (contentType.includes("application/json")) {
      const json = await response.json();
      Log.debug({ path, json, status }, "Response was application/json");
      if (!response.ok) {
        this.handleErrorResponse(response, method, path, json);
      }
      return responseSchema.parse(json);
    }
    if (contentType.includes("text/plain")) {
      const text = await response.text();
      Log.debug({ path, text, status }, "Response was text/plain");
      if (!response.ok) {
        this.handleErrorResponse(response, method, path, text);
      }
      return responseSchema.parse(text);
    }
    Log.debug({ path, status }, "Response had no body");
    if (!response.ok) {
      this.handleErrorResponse(response, method, path, null);
    }
    return responseSchema.parse(void 0);
  }
  /**
   * Retrieves comprehensive node information including version and configuration
   */
  async aboutNode() {
    return await this.request({
      path: NilDbEndpoint.v1.system.about,
      responseSchema: ReadAboutNodeResponse
    });
  }
  /**
   * Checks node health status
   */
  async healthCheck() {
    return await this.request({
      path: NilDbEndpoint.v1.system.health,
      responseSchema: NodeHealthCheckResponse
    });
  }
};

// src/nildb/builder-client.ts
var NilDbBuilderClientOptions = z10.object({
  ...NilDbBaseClientOptions.shape
});
var NilDbBuilderClient = class extends NilDbBaseClient {
  #options;
  constructor(options) {
    super(options);
    this.#options = options;
  }
  /**
   * Registers a new builder.
   */
  register(body) {
    return this.request({
      path: NilDbEndpoint.v1.builders.register,
      method: "POST",
      body,
      responseSchema: RegisterBuilderResponse
    });
  }
  /**
   * Retrieves the authenticated builder's profile information.
   */
  readProfile(token) {
    return this.request({
      path: NilDbEndpoint.v1.builders.me,
      token,
      responseSchema: ReadBuilderProfileResponse
    });
  }
  /**
   * Updates the authenticated builder's profile information.
   */
  updateProfile(token, body) {
    return this.request({
      path: NilDbEndpoint.v1.builders.me,
      method: "POST",
      body,
      token,
      responseSchema: UpdateBuilderProfileResponse
    });
  }
  /**
   * Deletes the authenticated builder and all associated resources.
   */
  deleteBuilder(token) {
    return this.request({
      path: NilDbEndpoint.v1.builders.me,
      method: "DELETE",
      token,
      responseSchema: DeleteBuilderResponse
    });
  }
  /**
   * Creates a new collection for data validation.
   */
  createCollection(token, body) {
    return this.request({
      path: NilDbEndpoint.v1.collections.root,
      method: "POST",
      body,
      token,
      responseSchema: CreateCollectionResponse
    });
  }
  /**
   * Lists all collections owned by the authenticated builder.
   */
  readCollections(token) {
    return this.request({
      path: NilDbEndpoint.v1.collections.root,
      method: "GET",
      token,
      responseSchema: ListCollectionsResponse
    });
  }
  /**
   * Deletes a collection by id and all associated data.
   */
  deleteCollection(token, collection) {
    return this.request({
      path: NilDbEndpoint.v1.collections.byId.replace(":id", collection),
      method: "DELETE",
      token,
      responseSchema: DeleteCollectionResponse
    });
  }
  /**
   * Retrieves a collection by id including metadata.
   */
  readCollection(token, collection) {
    return this.request({
      path: NilDbEndpoint.v1.collections.byId.replace(":id", collection),
      method: "GET",
      token,
      responseSchema: ReadCollectionMetadataResponse
    });
  }
  /**
   * Creates an index on a collection.
   */
  createCollectionIndex(token, collection, body) {
    return this.request({
      path: NilDbEndpoint.v1.collections.indexesById.replace(":id", collection),
      method: "POST",
      body,
      token,
      responseSchema: CreateCollectionIndexResponse
    });
  }
  /**
   * Drops an index from a collection.
   */
  dropCollectionIndex(token, collection, index) {
    return this.request({
      path: NilDbEndpoint.v1.collections.indexesByNameById.replace(":id", collection).replace(":name", index),
      method: "DELETE",
      token,
      responseSchema: DropCollectionIndexResponse
    });
  }
  /**
   * Lists all queries owned by the authenticated builder.
   */
  getQueries(token) {
    return this.request({
      path: NilDbEndpoint.v1.queries.root,
      token,
      responseSchema: ReadQueriesResponse
    });
  }
  /**
   * Retrieves a query by id.
   */
  getQuery(token, query) {
    return this.request({
      path: NilDbEndpoint.v1.queries.byId.replace(":id", query),
      token,
      responseSchema: ReadQueryResponse
    });
  }
  /**
   * Creates a new MongoDB aggregation query with variable substitution.
   */
  createQuery(token, body) {
    return this.request({
      path: NilDbEndpoint.v1.queries.root,
      method: "POST",
      body,
      token,
      responseSchema: CreateQueryResponse
    });
  }
  /**
   * Deletes a query by id.
   */
  deleteQuery(token, query) {
    return this.request({
      path: NilDbEndpoint.v1.queries.byId.replace(":id", query),
      method: "DELETE",
      token,
      responseSchema: DeleteQueryResponse
    });
  }
  /**
   * Executes a query with variable substitution.
   */
  runQuery(token, body) {
    return this.request({
      path: NilDbEndpoint.v1.queries.run,
      method: "POST",
      body,
      token,
      responseSchema: RunQueryResponse
    });
  }
  /**
   * Retrieves the status and results of a background query job.
   */
  readQueryRunResults(token, run) {
    return this.request({
      path: NilDbEndpoint.v1.queries.runById.replace(":id", run),
      token,
      responseSchema: ReadQueryRunByIdResponse
    });
  }
  /**
   * Uploads standard data records to a schema-validated collection.
   */
  createStandardData(token, body) {
    return this.request({
      path: NilDbEndpoint.v1.data.createStandard,
      method: "POST",
      body,
      token,
      responseSchema: CreateDataResponse
    });
  }
  /**
   * Searches for data matching the provided filter.
   */
  findData(token, body) {
    return this.request({
      path: NilDbEndpoint.v1.data.find,
      method: "POST",
      body,
      token,
      responseSchema: FindDataResponse
    });
  }
  /**
   * Updates data records matching the provided filter.
   */
  updateData(token, body) {
    return this.request({
      path: NilDbEndpoint.v1.data.update,
      method: "POST",
      body,
      token,
      responseSchema: UpdateDataResponse
    });
  }
  /**
   * Deletes data records matching the provided filter.
   */
  deleteData(token, body) {
    return this.request({
      path: NilDbEndpoint.v1.data.delete,
      method: "POST",
      body,
      token,
      responseSchema: DeleteDataResponse
    });
  }
  /**
   * Removes all data from a collection.
   */
  flushData(token, collection) {
    return this.request({
      path: NilDbEndpoint.v1.data.flushById.replace(":id", collection),
      method: "DELETE",
      token,
      responseSchema: FlushDataResponse
    });
  }
  /**
   * Retrieves the most recent data records from a collection.
   */
  tailData(token, collection, limit = 10) {
    return this.request({
      path: `${NilDbEndpoint.v1.data.tailById.replace(":id", collection)}?limit=${limit}`,
      method: "GET",
      token,
      responseSchema: TailDataResponse
    });
  }
};
async function createNilDbBuilderClient(baseUrl) {
  const response = await fetch(`${baseUrl}/about`);
  const body = await response.json();
  const validated = NilDbBuilderClientOptions.parse({
    about: body,
    baseUrl
  });
  return new NilDbBuilderClient(validated);
}

// src/builder.ts
var SecretVaultBuilderClient = class _SecretVaultBuilderClient extends SecretVaultBaseClient {
  /**
   * Creates and initializes a new SecretVaultBuilderClient instance.
   */
  static async from(options) {
    const { urls, keypair, blindfold } = options;
    Log.debug(
      {
        did: keypair.toDid().toString(),
        dbCount: urls.dbs.length,
        blindfold: !!blindfold
      },
      "Creating SecretVaultBuilderClient"
    );
    const payerBuilder = await new PayerBuilder().keypair(keypair).chainUrl(urls.chain).build();
    const nilauthClient = await NilauthClient.from(urls.auth, payerBuilder);
    const clientPromises = urls.dbs.map(
      (base) => createNilDbBuilderClient(base)
    );
    const clients = await Promise.all(clientPromises);
    let client;
    if (blindfold) {
      if ("key" in blindfold) {
        client = new _SecretVaultBuilderClient({
          clients,
          keypair,
          key: blindfold.key,
          nilauthClient
        });
      } else {
        const key = await toBlindfoldKey({
          ...blindfold,
          clusterSize: clients.length
        });
        client = new _SecretVaultBuilderClient({
          clients,
          keypair,
          key,
          nilauthClient
        });
      }
    } else {
      client = new _SecretVaultBuilderClient({
        clients,
        keypair,
        nilauthClient
      });
    }
    Log.info(
      {
        id: keypair.toDid().toString().slice(-8),
        nodes: clients.length,
        encryption: client._options.key?.constructor.name ?? "none"
      },
      "SecretVaultBuilderClient created"
    );
    return client;
  }
  #rootToken = null;
  #nilauthClient;
  constructor(options) {
    super(options);
    this.#nilauthClient = options.nilauthClient;
  }
  get rootToken() {
    if (!this.#rootToken) {
      throw new Error("`refreshRootToken` must be called first");
    }
    return this.#rootToken;
  }
  /**
   * Fetches a new root NUC token from the configured nilAuth server.
   */
  async refreshRootToken() {
    Log.debug("Refreshing root token");
    const { token } = await this.#nilauthClient.requestToken(
      this._options.keypair,
      "nildb"
    );
    this.#rootToken = token;
    Log.info({ builder: this.id }, "Root token refreshed");
  }
  /**
   * Checks subscription status by the builder's Did.
   */
  subscriptionStatus() {
    return this.#nilauthClient.subscriptionStatus(
      this.keypair.publicKey("hex"),
      "nildb"
    );
  }
  /**
   * Registers the builder with all nodes in the cluster.
   */
  async register(body) {
    const result = await executeOnCluster(this.nodes, (c) => c.register(body));
    Log.info({ builder: this.id }, "Builder registered");
    return result;
  }
  /**
   * Reads the builder's profile from the cluster.
   */
  async readProfile() {
    const resultsByNode = await executeOnCluster(this.nodes, (client) => {
      const token = this.mintRootInvocation({
        audience: client.id,
        command: NucCmd.nil.db.builders.read
      });
      return client.readProfile(token);
    });
    const result = processPlaintextResponse(resultsByNode);
    Log.info({ builder: this.id }, "Builder profile read");
    return result;
  }
  /**
   * Updates the builder's profile on all nodes.
   */
  async updateBuilderProfile(body) {
    const result = await executeOnCluster(this.nodes, (client) => {
      const token = this.mintRootInvocation({
        audience: client.id,
        command: NucCmd.nil.db.builders.update
      });
      return client.updateProfile(token, body);
    });
    Log.info(
      { builder: this.id, updateFields: Object.keys(body) },
      "Builder profile updated"
    );
    return result;
  }
  /**
   * Deletes the builder and associated resources from all nodes.
   */
  async deleteBuilder() {
    const result = await executeOnCluster(this.nodes, (client) => {
      const token = this.mintRootInvocation({
        audience: client.id,
        command: NucCmd.nil.db.builders.delete
      });
      return client.deleteBuilder(token);
    });
    Log.info({ builder: this.id }, "Builder deleted");
    return result;
  }
  /**
   * Creates a new collection on all nodes.
   */
  async createCollection(body) {
    const result = await executeOnCluster(this.nodes, (client) => {
      const token = this.mintRootInvocation({
        audience: client.id,
        command: NucCmd.nil.db.collections.create
      });
      return client.createCollection(token, body);
    });
    Log.info({ builder: this.id, collection: body.name }, "Collection created");
    return result;
  }
  /**
   * Reads a list of all collections from the cluster.
   */
  async readCollections() {
    const resultsByNode = await executeOnCluster(this.nodes, (client) => {
      const token = this.mintRootInvocation({
        audience: client.id,
        command: NucCmd.nil.db.collections.read
      });
      return client.readCollections(token);
    });
    const result = processPlaintextResponse(resultsByNode);
    Log.info(
      {
        builder: this.id,
        count: result.data?.length || 0
      },
      "Collections read"
    );
    return result;
  }
  /**
   * Reads the metadata for a single collection.
   */
  async readCollection(collection) {
    const resultsByNode = await executeOnCluster(this.nodes, (client) => {
      const token = this.mintRootInvocation({
        audience: client.id,
        command: NucCmd.nil.db.collections.read
      });
      return client.readCollection(token, collection);
    });
    const result = processPlaintextResponse(resultsByNode);
    Log.info({ builder: this.id, collection }, "Collection metadata read");
    return result;
  }
  /**
   * Deletes a collection its data from all nodes.
   */
  async deleteCollection(collection) {
    const result = await executeOnCluster(this.nodes, (client) => {
      const token = this.mintRootInvocation({
        audience: client.id,
        command: NucCmd.nil.db.collections.delete
      });
      return client.deleteCollection(token, collection);
    });
    Log.info({ builder: this.id, collection }, "Collection deleted");
    return result;
  }
  /**
   * Creates a new index on a collection.
   */
  async createCollectionIndex(collection, body) {
    const result = await executeOnCluster(this.nodes, (client) => {
      const token = this.mintRootInvocation({
        audience: client.id,
        command: NucCmd.nil.db.collections.update
      });
      return client.createCollectionIndex(token, collection, body);
    });
    Log.info(
      {
        builder: this.id,
        collection,
        name: body.name
      },
      "Collection index created"
    );
    return result;
  }
  /**
   * Drops an index from a collection.
   */
  async dropCollectionIndex(collection, index) {
    const result = await executeOnCluster(this.nodes, (client) => {
      const token = this.mintRootInvocation({
        audience: client.id,
        command: NucCmd.nil.db.collections.update
      });
      return client.dropCollectionIndex(token, collection, index);
    });
    Log.info(
      {
        builder: this.id,
        collection,
        index
      },
      "Collection index dropped"
    );
    return result;
  }
  /**
   * Creates one or more standard data documents in a collection.
   */
  async createStandardData(options) {
    const { body, delegation } = options;
    const { key, clients } = this._options;
    const nodePayloads = key ? await prepareConcealedRequest({ key, clients, body }) : preparePlaintextRequest({ clients, body });
    const result = await executeOnCluster(this.nodes, (client) => {
      let token = delegation;
      if (!token) {
        token = this.mintRootInvocation({
          audience: client.id,
          command: NucCmd.nil.db.data.create
        });
      }
      const id = Did.parse(client.id.toString());
      const payload = nodePayloads[id];
      return client.createStandardData(token, payload);
    });
    Log.info(
      {
        collection: body.collection,
        count: body.data.length,
        builder: this.id,
        isConcealed: !!key
      },
      "Data created"
    );
    return result;
  }
  /**
   * Retrieves a list of all saved queries.
   */
  async getQueries() {
    const result = await executeOnCluster(this.nodes, async (client) => {
      const token = this.mintRootInvocation({
        audience: client.id,
        command: NucCmd.nil.db.queries.read
      });
      return client.getQueries(token);
    });
    Log.info({ builder: this.id }, "Queries read");
    return result;
  }
  /**
   * Retrieves a single saved query by its id.
   */
  async getQuery(query) {
    const result = await executeOnCluster(this.nodes, async (client) => {
      const token = this.mintRootInvocation({
        audience: client.id,
        command: NucCmd.nil.db.queries.read
      });
      return client.getQuery(token, query);
    });
    Log.info({ query, builder: this.id }, "Query read");
    return result;
  }
  /**
   * Creates a new saved query on all nodes.
   */
  async createQuery(body) {
    const result = await executeOnCluster(this.nodes, async (client) => {
      const token = this.mintRootInvocation({
        audience: client.id,
        command: NucCmd.nil.db.queries.create
      });
      return client.createQuery(token, body);
    });
    Log.info(
      {
        builder: this.id,
        name: body.name,
        id: body._id,
        collection: body.collection
      },
      "Created query"
    );
    return result;
  }
  /**
   * Deletes a saved query from all nodes.
   */
  async deleteQuery(query) {
    const result = await executeOnCluster(this.nodes, async (client) => {
      const token = this.mintRootInvocation({
        audience: client.id,
        command: NucCmd.nil.db.queries.delete
      });
      return client.deleteQuery(token, query);
    });
    Log.info({ builder: this.id, query }, "Query deleted");
    return result;
  }
  /**
   * Starts a query execution job.
   */
  async runQuery(body) {
    const result = await executeOnCluster(this.nodes, async (client) => {
      const token = this.mintRootInvocation({
        audience: client.id,
        command: NucCmd.nil.db.queries.execute
      });
      return client.runQuery(token, body);
    });
    Log.info(
      {
        builder: this.id,
        query: body._id,
        run: Object.values(result)[0]?.data
      },
      "Started query run"
    );
    return result;
  }
  /**
   * Reads the results of a completed query run from each node.
   */
  readQueryRunResults(runs) {
    return executeOnCluster(this.nodes, async (client) => {
      const token = this.mintRootInvocation({
        audience: client.id,
        command: NucCmd.nil.db.queries.read
      });
      const id = Did.parse(client.id.toString());
      const run = runs[id];
      return client.readQueryRunResults(token, run);
    });
  }
  /**
   * Finds data in a collection, revealing concealed values if a key is configured.
   */
  async findData(body) {
    const resultsByNode = await executeOnCluster(this.nodes, (client) => {
      const token = this.mintRootInvocation({
        audience: client.id,
        command: NucCmd.nil.db.data.read
      });
      return client.findData(token, body);
    });
    const { key } = this._options;
    let result;
    if (key) {
      const data = await processConcealedListResponse({ key, resultsByNode });
      result = { data };
    } else {
      result = processPlaintextResponse(resultsByNode);
    }
    Log.info(
      {
        builder: this.id,
        collection: body.collection,
        count: result.data?.length || 0
      },
      "Data found"
    );
    return result;
  }
  /**
   * Updates documents in a collection, concealing the update payload if a key is configured.
   */
  async updateData(body) {
    const { key, clients } = this._options;
    let nodePayloads;
    if (key) {
      const concealedSetShares = await conceal(key, body.update);
      if (concealedSetShares.length !== clients.length) {
        throw new Error("Concealed shares count must match node count.");
      }
      const pairs = clients.map((client, index) => {
        const payload = {
          ...body,
          update: { $set: concealedSetShares[index] }
        };
        return [client.id.toString(), payload];
      });
      nodePayloads = Object.fromEntries(pairs);
    } else {
      nodePayloads = preparePlaintextRequest({ clients, body });
    }
    const result = await executeOnCluster(this.nodes, (client) => {
      const token = NucTokenBuilder.extending(this.rootToken).command(NucCmd.nil.db.data.update).audience(client.id).build(this.keypair.privateKey());
      const id = Did.parse(client.id.toString());
      return client.updateData(token, nodePayloads[id]);
    });
    Log.info(
      {
        builder: this.id,
        collection: body.collection,
        filter: body.filter
      },
      "Data updated"
    );
    return result;
  }
  /**
   * Deletes data from a collection based on a filter.
   */
  async deleteData(body) {
    const result = await executeOnCluster(this.nodes, async (client) => {
      const token = this.mintRootInvocation({
        audience: client.id,
        command: NucCmd.nil.db.data.delete
      });
      return client.deleteData(token, body);
    });
    Log.info(
      {
        builder: this.id,
        collection: body.collection,
        filter: body.filter
      },
      "Data deleted"
    );
    return result;
  }
  /**
   * Deletes all data from a collection.
   */
  async flushData(collection) {
    const result = await executeOnCluster(this.nodes, async (client) => {
      const token = this.mintRootInvocation({
        audience: client.id,
        command: NucCmd.nil.db.data.delete
      });
      return client.flushData(token, collection);
    });
    Log.info({ collection }, "Flushed data");
    return result;
  }
  /**
   * Reads the last N documents from a collection, revealing concealed values if a key is configured.
   */
  async tailData(collection, limit = 10) {
    Log.debug({ collection, limit }, "Tailing data");
    const resultsByNode = await executeOnCluster(this.nodes, (client) => {
      const token = this.mintRootInvocation({
        audience: client.id,
        command: NucCmd.nil.db.data.read
      });
      return client.tailData(token, collection, limit);
    });
    const { key } = this._options;
    let result;
    if (key) {
      const data = await processConcealedListResponse({ key, resultsByNode });
      result = { data };
    } else {
      result = processPlaintextResponse(resultsByNode);
    }
    Log.info({ collection, count: result.data?.length || 0 }, "Data tailed");
    return result;
  }
  mintRootInvocation(options) {
    return NucTokenBuilder.extending(this.rootToken).command(options.command).body(new InvocationBody({})).expiresAt(intoSecondsFromNow(60)).audience(options.audience).build(this.keypair.privateKey());
  }
};

// src/user.ts
import {
  InvocationBody as InvocationBody2,
  NucTokenBuilder as NucTokenBuilder2,
  NucTokenEnvelopeSchema
} from "@nillion/nuc";

// src/nildb/user-client.ts
import { z as z11 } from "zod/v4";
var NilDbUserClientOptions = z11.object({
  ...NilDbBaseClientOptions.shape
});
var NilDbUserClient = class extends NilDbBaseClient {
  #options;
  constructor(options) {
    super(options);
    this.#options = options;
  }
  /**
   * Retrieves the authenticated user's profile information.
   */
  readProfile(token) {
    return this.request({
      path: NilDbEndpoint.v1.users.me,
      token,
      responseSchema: ReadUserProfileResponse
    });
  }
  /**
   * Lists all data records owned by the authenticated user.
   */
  listDataReferences(token) {
    return this.request({
      path: NilDbEndpoint.v1.users.data.root,
      token,
      responseSchema: ListDataReferencesResponse
    });
  }
  /**
   * Create user-owned data in an owned collection
   */
  createOwnedData(token, body) {
    return this.request({
      path: NilDbEndpoint.v1.data.createOwned,
      method: "POST",
      token,
      body,
      responseSchema: CreateDataResponse
    });
  }
  /**
   * Retrieves user-owned data by collection and document id.
   */
  readData(token, params) {
    return this.request({
      path: NilDbEndpoint.v1.users.data.byId.replace(":collection", params.collection).replace(":document", params.document),
      token,
      responseSchema: ReadDataResponse
    });
  }
  /**
   * Deletes a user-owned data document.
   */
  deleteData(token, params) {
    return this.request({
      path: NilDbEndpoint.v1.users.data.byId.replace(":collection", params.collection).replace(":document", params.document),
      method: "DELETE",
      token,
      responseSchema: DeleteDocumentResponse
    });
  }
  /**
   * Grants access to user-owned data.
   */
  grantAccess(token, body) {
    return this.request({
      path: NilDbEndpoint.v1.users.data.acl.grant,
      method: "POST",
      body,
      token,
      responseSchema: GrantAccessToDataResponse
    });
  }
  /**
   * Removes access to user-owned data.
   */
  revokeAccess(token, body) {
    return this.request({
      path: NilDbEndpoint.v1.users.data.acl.revoke,
      method: "POST",
      body,
      token,
      responseSchema: RevokeAccessToDataResponse
    });
  }
};
async function createNilDbUserClient(baseUrl) {
  const response = await fetch(`${baseUrl}/about`);
  const about = await response.json();
  const validated = NilDbUserClientOptions.parse({
    about,
    baseUrl
  });
  return new NilDbUserClient(validated);
}

// src/user.ts
var SecretVaultUserClient = class _SecretVaultUserClient extends SecretVaultBaseClient {
  /**
   * Creates and initializes a new SecretVaultUserClient instance.
   */
  static async from(options) {
    const { baseUrls, keypair, blindfold } = options;
    const clientPromises = baseUrls.map((u) => createNilDbUserClient(u));
    const clients = await Promise.all(clientPromises);
    let client;
    if (blindfold) {
      if ("key" in blindfold) {
        client = new _SecretVaultUserClient({
          clients,
          keypair,
          key: blindfold.key
        });
      } else {
        const key = await toBlindfoldKey({
          ...blindfold,
          clusterSize: clients.length
        });
        client = new _SecretVaultUserClient({
          clients,
          keypair,
          key
        });
      }
    } else {
      client = new _SecretVaultUserClient({
        clients,
        keypair
      });
    }
    Log.info(
      {
        did: keypair.toDid().toString(),
        nodes: clients.length,
        encryption: client._options.key?.constructor.name ?? "none"
      },
      "SecretVaultUserClient created"
    );
    return client;
  }
  /**
   * Reads the user's profile information from the cluster.
   */
  async readProfile() {
    const resultsByNode = await executeOnCluster(this.nodes, (client) => {
      const token = this.mintInvocation({
        command: NucCmd.nil.db.users.root,
        audience: client.id
      });
      return client.readProfile(token);
    });
    const result = processPlaintextResponse(resultsByNode);
    Log.info({ user: this.id }, "User profile read");
    return result;
  }
  /**
   * Creates one or more data documents owned by the user.
   */
  async createData(delegation, body) {
    const { key, clients } = this._options;
    const nodePayloads = key ? await prepareConcealedRequest({ key, clients, body }) : preparePlaintextRequest({ clients, body });
    const result = await executeOnCluster(this.nodes, (client) => {
      const envelop = NucTokenEnvelopeSchema.parse(delegation);
      const token = NucTokenBuilder2.extending(envelop).audience(client.id).command(NucCmd.nil.db.data.create).expiresAt(intoSecondsFromNow(60)).body(new InvocationBody2({})).build(this.keypair.privateKey());
      const id = Did.parse(client.id.toString());
      const payload = nodePayloads[id];
      return client.createOwnedData(token, payload);
    });
    Log.info(
      {
        user: this.id,
        collection: body.collection,
        documents: body.data.length,
        concealed: !!key
      },
      "User data created"
    );
    return result;
  }
  /**
   * Lists references to all data documents owned by the user.
   */
  async listDataReferences() {
    const resultsByNode = await executeOnCluster(this.nodes, (client) => {
      const token = this.mintInvocation({
        command: NucCmd.nil.db.users.read,
        audience: client.id
      });
      return client.listDataReferences(token);
    });
    const result = processPlaintextResponse(resultsByNode);
    Log.info(
      { user: this.id, count: result.data?.length || 0 },
      "User data references listed"
    );
    return result;
  }
  /**
   * Reads a single data document, automatically revealing concealed values if a key is configured.
   */
  async readData(params) {
    const resultsByNode = await executeOnCluster(this.nodes, (client) => {
      const token = this.mintInvocation({
        command: NucCmd.nil.db.users.read,
        audience: client.id
      });
      return client.readData(token, params);
    });
    const { key } = this._options;
    let result;
    if (key) {
      const data = await processConcealedObjectResponse({
        key,
        resultsByNode
      });
      result = { data };
    } else {
      result = processPlaintextResponse(resultsByNode);
    }
    Log.info(
      {
        user: this.id,
        collection: params.collection,
        document: params.document
      },
      "User data read"
    );
    return result;
  }
  /**
   * Deletes a user-owned document from all nodes.
   */
  async deleteData(params) {
    const result = await executeOnCluster(this.nodes, (client) => {
      const token = this.mintInvocation({
        command: NucCmd.nil.db.users.delete,
        audience: client.id
      });
      return client.deleteData(token, params);
    });
    Log.info(
      {
        user: this.id,
        collection: params.collection,
        document: params.document
      },
      "User data deleted"
    );
    return result;
  }
  /**
   * Grants a given Did access to a given user-owned document.
   */
  async grantAccess(body) {
    const result = await executeOnCluster(this.nodes, (client) => {
      const token = this.mintInvocation({
        command: NucCmd.nil.db.users.update,
        audience: client.id
      });
      return client.grantAccess(token, body);
    });
    Log.info(
      {
        user: this.id,
        collection: body.collection,
        document: body.document,
        grantee: body.acl.grantee
      },
      "Data access granted"
    );
    return result;
  }
  /**
   * Revokes access for a given Did to the specified user-owned document.
   */
  async revokeAccess(body) {
    const result = await executeOnCluster(this.nodes, (client) => {
      const token = this.mintInvocation({
        command: NucCmd.nil.db.users.update,
        audience: client.id
      });
      return client.revokeAccess(token, body);
    });
    Log.info(
      {
        user: this.id,
        collection: body.collection,
        document: body.document,
        revokee: body.grantee
      },
      "Data access revoked"
    );
    return result;
  }
  mintInvocation(options) {
    const builder = NucTokenBuilder2.invocation({});
    return builder.command(options.command).subject(this.did).audience(options.audience).expiresAt(intoSecondsFromNow(60)).build(this.keypair.privateKey());
  }
};
export {
  Acl,
  AclDto,
  ApiErrorResponse,
  ApiSuccessResponse,
  ByIdRequestParams,
  CollectionDocumentDto,
  CollectionDocuments,
  CollectionIndexDto,
  CollectionMetadataDto,
  CreateCollectionIndexRequest,
  CreateCollectionIndexResponse,
  CreateCollectionRequest,
  CreateCollectionResponse,
  CreateDataResponse,
  CreateOwnedDataRequest,
  CreateQueryRequest,
  CreateQueryResponse,
  CreateStandardDataRequest,
  DataSchemaByIdRequestParams,
  DeleteBuilderResponse,
  DeleteCollectionRequestParams,
  DeleteCollectionResponse,
  DeleteDataRequest,
  DeleteDataResponse,
  DeleteDocumentRequestParams,
  DeleteDocumentResponse,
  DeleteQueryRequest,
  DeleteQueryResponse,
  Did,
  DropCollectionIndexParams,
  DropCollectionIndexResponse,
  DropDataResponse,
  FindDataRequest,
  FindDataResponse,
  FlushDataRequest,
  FlushDataResponse,
  GrantAccessToDataRequest,
  GrantAccessToDataResponse,
  ListCollectionsResponse,
  ListDataReferencesResponse,
  Name,
  NodeHealthCheckResponse,
  NucCmd,
  QueryVariableValidator,
  ReadAboutNodeResponse,
  ReadBuilderProfileResponse,
  ReadCollectionMetadataRequestParams,
  ReadCollectionMetadataResponse,
  ReadDataAccessResponse,
  ReadDataAclRequestParams,
  ReadDataRequestParams,
  ReadDataResponse,
  ReadQueriesResponse,
  ReadQueryResponse,
  ReadQueryRunByIdResponse,
  ReadUserProfileResponse,
  RegisterBuilderRequest,
  RegisterBuilderResponse,
  RevokeAccessToDataRequest,
  RevokeAccessToDataResponse,
  RunQueryRequest,
  RunQueryResponse,
  RunQueryResultStatus,
  SecretVaultBuilderClient,
  SecretVaultUserClient,
  TailDataRequestParams,
  TailDataRequestQuery,
  TailDataResponse,
  UpdateBuilderProfileRequest,
  UpdateBuilderProfileResponse,
  UpdateDataRequest,
  UpdateDataResponse,
  UpdateUserDataRequest,
  Uuid
};
